blog/2025-10-15-fargate-vs-managed-node-groups.mdx

```diff
@@ -0,0 +1,153 @@
+
+By AWS and Cloud Posse best practices, production-grade clusters should span three availability zones, with cluster-critical services distributed across them.
+
+However, during initial scheduling with **managed node groups**, Karpenter might spin up just one node large enough to fit all your add-on pods — even if they request three replicas with anti-affinity rules. Kubernetes will happily co-locate them all on that single node.
+
+Once they're running, those pods don't move automatically, even as the cluster grows. The result?
+
+**A deceptively healthy cluster with all your CoreDNS replicas living on the same node in one AZ — a single point of failure disguised as a distributed system.**
+
+While [topologySpreadConstraints](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/) can help encourage multi-AZ distribution, they don't guarantee it during the critical cluster bootstrap phase when Karpenter is creating its first nodes.
+
+## The Solution: A Minimal Managed Node Pool
+
+Our solution is simple:
+
+**Deploy a tiny managed node group — one node per availability zone — as part of your base cluster.**
+
+<FeatureList>
+  - This provides a home for cluster-critical add-ons during creation
+  - It ensures that CoreDNS, EBS CSI, and other vital components are naturally distributed across AZs
+  - It gives Karpenter a stable platform to run on
+  - And it eliminates the bootstrap deadlock problem entirely
+</FeatureList>
+
+You can even disable autoscaling for this node pool. One node per AZ is enough.
+
+Think of it as your cluster's heartbeat — steady, predictable, and inexpensive.
+
+## Additional Fargate Constraints
+
+Beyond the HA challenges, [Fargate has architectural constraints](https://docs.aws.amazon.com/eks/latest/userguide/fargate-pod-configuration.html) that can affect cluster add-ons:
+
+<FeatureList>
+  - Each Fargate pod runs on its own isolated compute resource (one pod per node)
+  - No dynamic persistent volume provisioning
+  - Fixed CPU and memory configurations with coarse granularity
+  - 256 MB memory overhead for Kubernetes components
+</FeatureList>
+
+While these constraints don't necessarily prevent Fargate from working, they add complexity when running cluster-critical infrastructure that needs precise resource allocation and high availability guarantees.
+
+## Cost and Flexibility
+
+Fargate offers convenience, but at a premium. A pod requesting 2 vCPUs and 4 GiB of memory costs about **$0.098/hour**, compared to **$0.076/hour** for an equivalent EC2 c6a.large instance.
+
+And because [Fargate bills in coarse increments](https://docs.aws.amazon.com/eks/latest/userguide/fargate-pod-configuration.html), you often overpay for partial capacity.
+
+By contrast, a managed node group gives you flexibility:
+
+<FeatureList>
+  - Use Graviton-based instances (c7g.medium) to cut costs nearly in half
+  - Mix On-Demand nodes for reliability and Spot nodes (via Karpenter) for efficiency
```

Comment:
Surely the MNGs should have on-demand nodes only for stability, or is running Spot on MNGs also part of the recommendation?